src/common/btree.c

```diff
@@ -1881,13 +1881,10 @@ btr_insert(struct btr_context *tcx, d_iov_t *key, d_iov_t *val)
 		rc = btr_root_start(tcx, rec);
 		if (rc != 0) {
 			D_DEBUG(DB_TRACE, "Failed to start the tree: %d\n", rc);
-			goto failed;
+			return rc;
 		}
 	}
 	return 0;
- failed:
-	btr_rec_free(tcx, rec, NULL);
-	return rc;
```

Comment:
I think this kind of error cleanup is still useful when dbtree is on non-transactional vmem class, right?

Reply:
An excellent point.   I'm not sure we handle this very well

Reply:
I agree that it may be useful for VM based btree case, but only btr_rec_free() is not enough, because the @rec may be not the new created record, the split logic may have changed it. Currently, we have depends on PMDK transaction mechanism to undo related modifications, otherwise, the rollback will be very complex.

Reply:
In theory, the explicitly rollback for undo btr_insert() may fail, we do not have perfect solution for that unless related DAOS transaction can rollback automatically, such as aborting the PMDK transaction.

Reply:
I think the only cause to fail tree split is ENOSPC (ENOMEM), so if we preserve enough tree nodes before splitting, the error cleanup will be much more easier?

But I'm totally ok with this temporary fix.

Reply:
I am not sure about the reason for btr_insert failure, no description about the error# in the Jira ticket, and I cannot reproduce it manually. But I have to say that explicit rollback for b+tree split/renbalance is some complex against current implementation. So using PMDK undo logs is the most clean way. But I agree that if the b+tree in VM based, then we need other solution.

Reply:
I agree the failure handling is flawed, however, because @rec is a input parameter, should we just do this?
btr_node_split_and_insert(..., struct btr_record *rec)
{
          struct btr_recrod rec_tmp = *rec;

then we can safely overwrite members of rec_tmp and use it as input parameter for recursive functions?

Reply:
I do not think so, because if some failure happened during the btree split, the btree may become unbalance and in some intermediateness. If without PMDK undo logs, only releasing the new created @rec will help very limited. On the other hand, the new allocated @rec may have already been filled into some slot in the btree, if we release such @rec without PMDK rollback, then there may be some garbage information left in the btree.

Reply:
OK, let's land it for now, I will think it further